src/client.ts

```diff
@@ -117,41 +144,117 @@ export class OpenFeatureClient implements Client {
       options: FlagEvaluationOptions | undefined
     ) => Promise<ResolutionDetails<T>>,
     defaultValue: T,
-    context: EvaluationContext = {},
+    flagType: FlagType,
+    invocationContext: EvaluationContext = {},
     options: FlagEvaluationOptions = {}
   ): Promise<EvaluationDetails<T>> {
     // merge global, client, and evaluation context
-    const mergedContext = {
-      ...this.api.context,
+
+    const allHooks = [...OpenFeature.hooks, ...this.hooks, ...(options.hooks || [])];
+    const allHooksReversed = [...allHooks].reverse();
+
+    // merge global and client contexts
+    const globalAndClientContext = {
+      ...OpenFeature.context,
       ...this.context,
-      ...context,
+    };
+
+    // this reference cannot change during the course of evaluation
+    // it may be used as a key in WeakMaps
+    const hookContext: Readonly<HookContext> = {
+      flagKey,
+      defaultValue,
+      flagType,
+      client: this,
+      provider: OpenFeature.provider,
+      context: globalAndClientContext,
     };
 
     try {
-      // if a transformer is defined, run it to prepare the context.
+      const mergedHookContext = await this.beforeHooks(allHooks, hookContext, options);
+
+      // merge context in order: global, client, hook, invocation
+      const mergedContext = {
+        ...mergedHookContext,
+        ...invocationContext,
+      };
```

Comment:
Since the spec dictates invocation context "wins" over any hooks, I merge invocation context attributes after merging that from before hooks. I think doing it any other way would be inconsistent and confusing.

@justinabrahms  @beeme1mr

Reply:
That seems correct to me. Here's my implementation: 

https://github.com/open-feature/java-sdk/blob/main/lib/src/main/java/dev/openfeature/javasdk/OpenFeatureClient.java#L61

And here is where we update the before hook context: https://github.com/open-feature/java-sdk/blob/main/lib/src/main/java/dev/openfeature/javasdk/OpenFeatureClient.java#L121